dot/parachain/grid/grid_topology.go

```diff
@@ -0,0 +1,346 @@
+	PeersRow map[peer.ID]struct{}
+	PeersCol map[peer.ID]struct{}
+
+	ValidatorIndicesRow map[parachaintypes.ValidatorIndex]struct{}
+	ValidatorIndicesCol map[parachaintypes.ValidatorIndex]struct{}
+}
+
+func NewEmptyGridNeighbors() *GridNeighbors {
+	return &GridNeighbors{
+		PeersRow:            make(map[peer.ID]struct{}),
+		PeersCol:            make(map[peer.ID]struct{}),
+		ValidatorIndicesRow: make(map[parachaintypes.ValidatorIndex]struct{}),
+		ValidatorIndicesCol: make(map[parachaintypes.ValidatorIndex]struct{}),
+	}
+}
+
+// RequiredRoutingByIndex Given the originator of a message as a validator index, indicates the part of the topology
+func (gn *GridNeighbors) RequiredRoutingByIndex(origin parachaintypes.ValidatorIndex, local bool) RequiredRouting {
+	if local {
+		return RequiredRoutingGridXY
+	}
+	_, x := gn.ValidatorIndicesRow[origin]
+	_, y := gn.ValidatorIndicesCol[origin]
+
+	if x && y {
+		// If all works correctly origin peer can't be in both rows and columns. But we leave it anyways
+		return RequiredRoutingGridXY
+	}
+	if !(x || y) {
+		return RequiredRoutingNone
+	}
+	if x && !y {
+		return RequiredRoutingGridY
+	}
+	return RequiredRoutingGridX
+}
+
+// ShouldRouteToPeer indicates does peer should receive a message based on GridTopology and Routing strategy
+func (gn *GridNeighbors) ShouldRouteToPeer(routing RequiredRouting, peer peer.ID) bool {
+	switch routing {
+	case RequiredRoutingAll:
+		return true
+	case RequiredRoutingNone, PendingTopology:
+		return false
+	case RequiredRoutingGridXY:
+		_, x := gn.PeersRow[peer]
+		_, y := gn.PeersCol[peer]
+		return x || y
+	case RequiredRoutingGridX:
+		_, x := gn.PeersRow[peer]
+		return x
+	case RequiredRoutingGridY:
+		_, y := gn.PeersCol[peer]
+		return y
+	default:
+		// No way we get here
+		return false
+	}
+}
+
+// PeersDiff returns a differents between two GridNeighbors
+func (gn *GridNeighbors) PeersDiff(other *GridNeighbors) []peer.ID {
+	diff := make([]peer.ID, 0)
+	for p := range gn.PeersRow {
+		_, inRows := other.PeersRow[p]
+		_, inCols := other.PeersCol[p]
+		if !inRows && !inCols {
+			diff = append(diff, p)
+		}
+	}
+	return diff
+}
+
+func (gn *GridNeighbors) Len() int {
+	return len(gn.PeersRow) + len(gn.PeersCol)
+}
+
+type RequiredRouting uint
+
+const (
+	PendingTopology    RequiredRouting = iota
+	RequiredRoutingAll                 // All peers in the grid
+	RequiredRoutingGridXY
+	RequiredRoutingGridX
+	RequiredRoutingGridY
+	RequiredRoutingNone
+)
+
+// MatrixNeighbors holds the row and column neighbors of a given index in a matrix.
+type MatrixNeighbors struct {
+	RowNeighbors    []uint
+	ColumnNeighbors []uint
+}
+
+// CalculateMatrixNeighbors computes the row and column neighbors of valIndex in a matrix of given length.
+// e.g. for size 11 the matrix would be
+//
+// 0  1  2
+// 3  4  5
+// 6  7  8
+// 9 10
+//
+// and for index 10, the neighbors would be 1, 4, 7, 9
+func CalculateMatrixNeighbors(valIndex, length uint) (*MatrixNeighbors, error) {
```

Comment:
```suggestion
func calculateMatrixNeighbors(valIndex, length uint) (*MatrixNeighbors, error) {
```

if the function is only used under the package we can change it private